File: Sep12/Lecture7Examples/Example27/WeatherFragment.java
package com.varunshankar.example27;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.os.HandlerCompat;
import androidx.fragment.app.Fragment;

import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import org.json.JSONException;

import java.lang.ref.WeakReference;
import java.net.URL;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


public class WeatherFragment extends Fragment implements View.OnClickListener{

    private EditText mEtLocation;
    private TextView mTvTemp;
    private TextView mTvPress;
    private TextView mTvHum;
    private WeatherData mWeatherData;
    private Button mBtSubmit;
    private static FetchWeatherTask mFetchWeatherTask = new FetchWeatherTask();

    public WeatherFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_weather, container, false);
        //Get the edit text and all the text views
        mEtLocation = (EditText) view.findViewById(R.id.et_location);
        mTvTemp = (TextView) view.findViewById(R.id.tv_temp);
        mTvPress = (TextView) view.findViewById(R.id.tv_pressure);
        mTvHum = (TextView) view.findViewById(R.id.tv_humidity);
        if(savedInstanceState!=null) {
            String temp = savedInstanceState.getString("tvTemp");
            String hum = savedInstanceState.getString("tvHum");
            String press = savedInstanceState.getString("tvPress");
            if (temp != null)
                mTvTemp.setText(""+temp);
            if (hum != null)
                mTvHum.setText(""+hum);
            if (press != null)
                mTvPress.setText(""+press);
        }
        mFetchWeatherTask.setWeakReference(this); //make sure we're always pointing to current version of fragment
        mBtSubmit = (Button) view.findViewById(R.id.button_submit);
        mBtSubmit.setOnClickListener(this);
        return view;
    }
    @Override
    public void onClick(View view) {
        switch(view.getId()){
            case R.id.button_submit:{
                //Get the string from the edit text and sanitize the input
                String inputFromEt = mEtLocation.getText().toString().replace(' ','&');
                loadWeatherData(inputFromEt);
            }
            break;
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("tvTemp",mTvTemp.getText().toString());
        outState.putString("tvHum",mTvHum.getText().toString());
        outState.putString("tvPress",mTvPress.getText().toString());
    }

    private void loadWeatherData(String location){
      mFetchWeatherTask.execute(this,location);
    }
    private static class FetchWeatherTask{
        WeakReference<WeatherFragment> weatherFragmentWeakReference;
        private ExecutorService executorService = Executors.newSingleThreadExecutor();
        private Handler mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper());

        public void setWeakReference(WeatherFragment ref)
        {
            weatherFragmentWeakReference = new WeakReference<WeatherFragment>(ref);
        }
        public void execute(WeatherFragment ref, String location){

            executorService.execute(new Runnable(){
                @Override
                public void run(){
                    String jsonWeatherData;
                    URL weatherDataURL = NetworkUtils.buildURLFromString(location);
                    jsonWeatherData = null;
                    try{
                        jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL);
                        postToMainThread(jsonWeatherData);
                    }catch(Exception e){
                        e.printStackTrace();
                    }
                }
            });
        }

        private void postToMainThread(String jsonWeatherData)
        {
            WeatherFragment localRef = weatherFragmentWeakReference.get();
            mainThreadHandler.post(new Runnable() {
                @Override
                public void run() {
                    if (jsonWeatherData != null) {
                        try {
                            localRef.mWeatherData = JSONWeatherUtils.getWeatherData(jsonWeatherData);
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                        if (localRef.mWeatherData != null) {
                            localRef.mTvTemp.setText("" + Math.round(localRef.mWeatherData.getTemperature().getTemp() - 273.15) + " C");
                            localRef.mTvHum.setText("" + localRef.mWeatherData.getCurrentCondition().getHumidity() + "%");
                            localRef.mTvPress.setText("" + localRef.mWeatherData.getCurrentCondition().getPressure() + " hPa");
                        }
                    }
                }
            });
        }
    }
}

File: Sep12/Example27/app/src/main/java/com/varunshankar/example27/WeatherFragment.kt
package com.varunshankar.example27

import com.varunshankar.example27.NetworkUtils.buildURLFromString
import com.varunshankar.example27.NetworkUtils.getDataFromURL
import android.widget.EditText
import android.widget.TextView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import androidx.core.os.HandlerCompat
import android.os.Looper
import android.view.View
import android.widget.Button
import androidx.fragment.app.Fragment
import org.json.JSONException
import java.lang.Exception
import java.lang.ref.WeakReference
import java.util.concurrent.Executors

class WeatherFragment : Fragment(), View.OnClickListener {
    private var mEtLocation: EditText? = null
    private var mTvTemp: TextView? = null
    private var mTvPress: TextView? = null
    private var mTvHum: TextView? = null
    private var mWeatherData: WeatherData? = null
    private var mBtSubmit: Button? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_weather, container, false)
        //Get the edit text and all the text views
        mEtLocation = view.findViewById<View>(R.id.et_location) as EditText
        mTvTemp = view.findViewById<View>(R.id.tv_temp) as TextView
        mTvPress = view.findViewById<View>(R.id.tv_pressure) as TextView
        mTvHum = view.findViewById<View>(R.id.tv_humidity) as TextView
        if (savedInstanceState != null) {
            val temp = savedInstanceState.getString("tvTemp")
            val hum = savedInstanceState.getString("tvHum")
            val press = savedInstanceState.getString("tvPress")
            if (temp != null) mTvTemp!!.text = "" + temp
            if (hum != null) mTvHum!!.text = "" + hum
            if (press != null) mTvPress!!.text = "" + press
        }
        mFetchWeatherTask.setWeakReference(this) //make sure we're always pointing to current version of fragment
        mBtSubmit = view.findViewById<View>(R.id.button_submit) as Button
        mBtSubmit!!.setOnClickListener(this)
        return view
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.button_submit -> {

                //Get the string from the edit text and sanitize the input
                val inputFromEt = mEtLocation!!.text.toString().replace(' ', '&')
                loadWeatherData(inputFromEt)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("tvTemp", mTvTemp!!.text.toString())
        outState.putString("tvHum", mTvHum!!.text.toString())
        outState.putString("tvPress", mTvPress!!.text.toString())
    }

    private fun loadWeatherData(location: String) {
        mFetchWeatherTask.execute(location)
    }

    private class FetchWeatherTask {
        var weatherFragmentWeakReference: WeakReference<WeatherFragment>? = null
        private val executorService = Executors.newSingleThreadExecutor()
        private val mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper())
        fun setWeakReference(ref: WeatherFragment) {
            weatherFragmentWeakReference = WeakReference(ref)
        }

        fun execute(location: String?) {
            executorService.execute {
                var jsonWeatherData: String?
                val weatherDataURL = buildURLFromString(location!!)
                jsonWeatherData = null
                try {
                    jsonWeatherData = getDataFromURL(weatherDataURL!!)
                    postToMainThread(jsonWeatherData)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        private fun postToMainThread(jsonWeatherData: String?) {
            val localRef = weatherFragmentWeakReference!!.get()
            mainThreadHandler.post {
                if (jsonWeatherData != null) {
                    try {
                        localRef!!.mWeatherData = JSONWeatherUtils.getWeatherData(jsonWeatherData)
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                    if (localRef!!.mWeatherData != null) {
                        localRef.mTvTemp!!.text =
                            "" + Math.round(localRef.mWeatherData!!.temperature.temp - 273.15) + " C"
                        localRef.mTvHum!!.text =
                            "" + localRef.mWeatherData!!.currentCondition.humidity + "%"
                        localRef.mTvPress!!.text =
                            "" + localRef.mWeatherData!!.currentCondition.pressure + " hPa"
                    }
                }
            }
        }
    }

    companion object {
        private val mFetchWeatherTask = FetchWeatherTask()
    }
}